server/data/events.ts

```diff
@@ -0,0 +1,224 @@
+                "applications.$.events": newEvent
+                }
+            }
+        );
+        if (updateInfo.WriteResult.nModified === 0) throw new Error('Could not add event');
+
+        return exports.default.readById(cycleId, applicationId, newEvent._id);
+    },
+
+    /**
+     * Update event
+     * @param {ObjectId} cycleId 
+     * @param {ObjectId} applicationId 
+     * @param {ObjectId} eventId
+     * @param {object} eventObject can contain: {title: string, date: Date, location: string}
+     */
+    update: async (cycleId: ObjectId, applicationId: ObjectId, eventId: ObjectId, eventObject: object): Promise<object> => {
+
+
+        const cyclesCollection: any = await collections.cycles();
+        const cycle = await cyclesCollection.findOne({_id: new ObjectId(cycleId)});
+        if(cycle === null) throw new Error('Could not find cycle');
+
+        //find the application
+        let application = null;
+        for(let element of cycle.applications){
+            if(element._id.toString() === applicationId.toString()){
+                application = element;
+                break;
+            } 
+        }
+        if(application === null) throw new Error('Could not find application');
+
+        //find the event
+        let updateEvent = null;
+        for(let element of application.events){
+            if(element._id.toString() === eventId.toString()){
+                updateEvent = element
+            }
+        }
+        if(updateEvent === null) throw new Error('Could not find event');
+
+
+        //update the event
+        for(let attribute in eventObject){
+            if(attribute in updateEvent){
+                updateEvent[attribute] = eventObject[attribute];
+            } else {
+                throw new Error('Invalid event attribute');
+            }
+        }
+
+        //update the events list
+        let events = [];
+        for(let element of application.events){
+            if(element._id.toString() === eventId.toString()){
+                events.push(updateEvent);
+            }
+            events.push(element);
+        }
+
+        //update mongodb with new events list
+        let updateInfo = await cyclesCollection.update(
+            {
+            _id: new ObjectId(cycleId),
+            "applications._id": new ObjectId(applicationId)
+            }, 
+            {$set: {
+                "applications.$.events": events
+                }
+            }
+        );
+        if (updateInfo.WriteResult.nModified === 0) throw new Error('Could not update event');
+
+        return exports.default.readById(cycleId, applicationId, eventId);
+    },
+
+    /**
+     * Delete event
+     * @param {ObjectId} cycleId 
+     * @param {ObjectId} applicationId 
+     * @param {ObjectId} eventId 
+     * @returns true if the event was deleted, throws otherwise
+     */
+    delete: async (cycleId: ObjectId, applicationId: ObjectId, eventId: ObjectId): Promise<boolean> => {
```

Comment:
same series of comments
- error check
- if you can try to reuse code that's already been written